cli.js

```diff
@@ -1,20 +1,128 @@
 #!/usr/bin/env node
-'use strict'
-const spawnSync = require('child_process').spawnSync
-const COMPILER_PATH = require('./index.js').compiler.COMPILER_PATH
-const result = spawnSync(
-  'java',
-  ['-jar', COMPILER_PATH].concat(process.argv.slice(2)),
-  {stdio: 'inherit'}
-)
-if (result.error) {
-  if (result.error.code === 'ENOENT') {
-    console.error('Could not find "java" in your PATH.')
-  } else {
-    console.error(result.error.message)
-  }
-  process.exit(1)
+'use strict';
+const fs = require('fs');
+const {getNativeImagePath, getFirstSupportedPlatform, parseCliFlags} = require('./lib/utils');
+
+const compilerFlags = parseCliFlags(process.argv.slice(2));
+let platform;
+if (compilerFlags.platform) {
+  platform = compilerFlags.platform;
+  delete compilerFlags.platform;
 } else {
-  process.exit(result.status)
+  platform = getFirstSupportedPlatform(['native', 'java', 'javascript']);
 }
 
+if (platform !== 'javascript') {
+  const Compiler = require('./lib/node/closure-compiler');
+  let args = process.argv.slice(2);
+  for (let i = 0; i < args.length; i++) {
+    if (/^--platform/.test(args[i])) {
+      let delCount = 1;
+      if (args[i].indexOf('=') < 0 && args.length > i + 1) {
+        delCount++;
+      }
+      args.splice(i, delCount);
+      break;
+    }
+  }
+
+  const compiler = new Compiler(args);
+
+  compiler.spawnOptions = {stdio: 'inherit'};
+  if (platform === 'native') {
+    compiler.JAR_PATH = null;
+    compiler.javaPath = getNativeImagePath();
+  }
+
+  compiler.run((exitCode) => {
+    process.exit(exitCode);
+  });
+} else {
+  if (compilerFlags.help === true) {
+    console.log('Sample usage: --compilation_level (-O) VAL --externs VAL --js VAL --js_output_file VAL --warning_level (-W) [QUIET | DEFAULT | VERBOSE]');
+    console.log('See https://github.com/google/closure-compiler/wiki/Flags-and-Options for the full list of flags');
+    process.exit(0);
+  }
+
+  if (compilerFlags.version === true) {
+    const {version} = require('./package.json');
+    console.log(`Version: ${version}`);
+    process.exit(0);
+  }
+
+  let inputFilePromises = [];
+  let waitOnStdIn = true;
```

Comment:
Have to run.  Sending what I have for review right now.  Will give the promise and event handling code a close look when I resume.